duktape/src/main/jni/DuktapeContext.cpp

```diff
@@ -141,39 +150,61 @@ jstring DuktapeContext::evaluate(JNIEnv *env, jstring code, jstring fname) {
   return result;
 }
 
-void DuktapeContext::bindInstance(JNIEnv *env, jstring instance, jobject object,
-                                  jobjectArray methods) {
+void DuktapeContext::bindInstance(JNIEnv *env, jstring name, jobject object, jobjectArray methods) {
   duk_push_global_object(m_context);
+  const JString instanceName(env, name);
+  if (duk_has_prop_string(m_context, -1, instanceName)) {
+    duk_pop(m_context);
+    const jclass illegalArgumentException = env->FindClass("java/lang/IllegalArgumentException");
+    const std::string message = "A global object called " + instanceName.str() + " already exists";
+    env->ThrowNew(illegalArgumentException, message.c_str());
+    return;
+  }
   const duk_idx_t objIndex = duk_require_normalize_index(m_context, duk_push_object(m_context));
 
   // Keep a reference in JavaScript to the object being bound.
   duk_push_pointer(m_context, env->NewGlobalRef(object));
   duk_put_prop_string(m_context, objIndex, JAVA_THIS_PROP_NAME);
+
+  // Hook up a finalizer to decrement the refcount and clean up our JavaMethods.
   duk_push_c_function(m_context, javaObjectFinalizer, 1);
   duk_set_finalizer(m_context, objIndex);
 
   const jsize numMethods = env->GetArrayLength(methods);
   for (jsize i = 0; i < numMethods; ++i) {
     jobject method = env->GetObjectArrayElement(methods, i);
-    JavaMethod* javaMethod = new JavaMethod(env, method);
+
+    const jmethodID getName =
+        env->GetMethodID(env->GetObjectClass(method), "getName", "()Ljava/lang/String;");
+    const JString methodName(env, static_cast<jstring>(env->CallObjectMethod(method, getName)));
+
+    std::unique_ptr<JavaMethod> javaMethod;
+    try {
+      javaMethod.reset(new JavaMethod(env, method));
+    } catch (std::invalid_argument& invalidArgument) {
```

Comment:
this one surprises me though . . . why try/catch to test the argument rather than using an if? In Java code catching an IllegalArgumentException is almost always a bad idea